qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -410,43 +414,122 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
+                    if isinstance(decomposition, QuantumCircuit):
+                        decomposition = circuit_to_dag(decomposition, copy_operations=False)
+
+                    if isinstance(decomposition, DAGCircuit):
+                        # Compute the qubits on which to apply the decomposition
+                        # (including auxiliary qubits).
+                        num_ancillas_used = decomposition.num_qubits() - len(node.qargs)
+                        if not use_ancillas or num_ancillas_used == 0:
+                            extended_qargs = node.qargs
+                        elif num_ancillas_used <= len(clean_ancillas_available):
+                            extended_qargs = (
+                                list(node.qargs) + clean_ancillas_available[0:num_ancillas_used]
+                            )
+                        else:
+                            extended_qargs = (
+                                list(node.qargs)
+                                + clean_ancillas_available
+                                + dirty_ancillas_available[
+                                    0 : num_ancillas_used - len(clean_ancillas_available)
+                                ]
+                            )
+
+                        for decomposition_node in decomposition.op_nodes():
+                            q_indices = [
+                                decomposition.find_bit(q).index for q in decomposition_node.qargs
+                            ]
+                            c_indices = [
+                                decomposition.find_bit(q).index for q in decomposition_node.cargs
+                            ]
+                            new_dag.apply_operation_back(
+                                decomposition_node.op,
+                                [extended_qargs[i] for i in q_indices],
+                                [node.cargs[i] for i in c_indices],
+                                check=True,
+                            )
+                        new_dag.global_phase += decomposition.global_phase
+
+                    elif isinstance(decomposition, Operation):
+                        # As of now, HighLevelSynthesis can only produce an Operation when synthesizing
+                        # annotated operations (for example, a singly-controlled CX-gate may result in a
+                        # CCX-gate). There should be no case when this Operation would have more qubits
+                        # than the original operation.
+                        if decomposition.num_qubits != len(node.qargs):
+                            raise TranspilerError(
+                                "HighLevelSynthesis reached an unexpected result."
+                            )
+                        new_dag.apply_operation_back(decomposition, node.qargs, node.cargs)
+
+            # update the list of clean ancilla qubits
+            clean_ancillas = [q for q in clean_ancillas if q not in node.qargs]
```

Comment:
We should add re-using clean ancillas in the future. One option would be to let the synthesis method report what it does with ancilla qubits, e.g. how many of which kind it requires, and how they are returned.

Reply:
I agree! We can probably use circuit's `metadata` for this (and any additional information that the synthesis method needs to return).